client-app/src/desktop/tabs/other/routing/AdvancedRoutingPanel.tsx

```diff
@@ -0,0 +1,171 @@
+                            {value: 'company|asc', label: 'Company Name (Asc)'},
+                            {value: 'city|desc', label: 'City (Desc)'},
+                            {value: 'city|asc', label: 'City (Asc)'},
+                            {value: 'trade_date|desc', label: 'Trade Date (Desc)'},
+                            {value: 'trade_date|asc', label: 'Trade Date (Asc)'},
+                            {value: null, label: 'None'}
+                        ]
+                    }),
+                    filler(),
+                    gridCountLabel({unit: 'companies'}),
+                    vbox({})
+                ]
+            })
+        });
+    }
+});
+
+class AdvancedRoutingPanelModel extends HoistModel {
+    @observable groupBy = null;
+    @observable sortBy = null;
+
+    constructor() {
+        super();
+        makeObservable(this);
+
+        this.addReaction({
+            track: () => XH.routerState.params,
+            run: () => this.parseRouteParams(),
+            fireImmediately: true
+        });
+
+        this.addReaction({
+            track: () => [this.groupBy, this.sortBy, this.gridModel.selectedRecord?.id],
+            run: () => this.updateRoute(),
+            fireImmediately: true
+        });
+    }
+
+    gridModel = new GridModel({
+        columns: [
+            {field: 'id', flex: 0},
+            {field: 'company', flex: 1},
+            {field: 'city', flex: 1},
+            {field: 'trade_date', flex: 1}
+        ]
+    });
+
+    @action
+    private setGroupBy(groupBy: string) {
+        this.groupBy = groupBy;
+
+        // Always select first when regrouping.
+        const groupByArr = groupBy ? groupBy.split(',') : [];
+        this.gridModel.setGroupBy(groupByArr);
+    }
+
+    @action
+    private setSortBy(sortBy: string) {
+        this.sortBy = sortBy;
+
+        // Always select first when resorting.
+        const sortByArr = sortBy ? sortBy.split(',') : [];
+        this.gridModel.setSortBy(sortByArr);
+    }
+
+    @action
+    private async setSelected(recordId: string | number) {
+        await this.gridModel.selectAsync(Number(recordId));
+    }
+
+    @action
+    private async parseRouteParams() {
+        const advParam = XH.routerState.params.routeParam;
+        if (!advParam) return;
+        const decodedParam = atob(advParam);
+        const {groupBy, sortBy, selectedId} = JSON.parse(decodedParam);
+        if (groupBy) this.setGroupBy(groupBy);
+        if (sortBy) this.setSortBy(sortBy);
+        if (selectedId) await this.setSelected(selectedId);
+    }
+
+    @action
+    private updateRoute() {
+        if (!XH.routerState.name.startsWith('default.other.advancedRouting')) return;
+        const {groupBy, sortBy} = this;
+        const selectedId = this.gridModel.selectedRecord?.id;
+        const routeParam = btoa(JSON.stringify({groupBy, sortBy, selectedId}));
+        XH.navigate('default.other.advancedRouting.routeParam', {routeParam});
```

Comment:
Probably want `{replace: true}` as the third arg so avoid extra history entries

Reply:
Implemented this selectively -- will replace when a new record is selected in the grid, but will navigate when grouping or sorting is changed.